### DISCUSSION TITLE
Sample breakout game has poor performance

### DISCUSSION THREAD

**Discussion by MorganeLeCaignec** [excaliburjs/Excalibur#2304] [Q&A]
Hello !
I thought I'd try this out to see if I liked it, as I'm interested in making simple games with TS.
My plan was to use Vite, and Vue on top of the excalibur app to handle UI, possibly to add complex and highly interactive UI elements (I'd also want to try integrating an excalibur game inside of a more "normal" webpage).

I made a simple little test with a Vue 3 + Vite + TS template, added excalibur and pasted the code from the sample breakout game (https://github.com/excaliburjs/sample-breakout/blob/main/src/main.ts#L17) in my main.ts.
I don't initialize/start the Vue App, only testing excalibur for now, and I serve with Vite in dev mode.

It seems to have pretty poor performances on both chrome and firefox. Even with browser devtools and excalibur-devtools off, it will not be able to maintain a constant 60 fps, constantly dropping to 50, 40, and will have seemingly random drops to 30 or lower even when the ball is just moving with no collision happening. Regularly the frames will drop at the wrong time and the ball will phase  or half-phase through the paddle or the screen walls. For some reason on firefox with debug toggled, the performances are abysmal, hovering around 30 FPS at best.

On both chrome and firefox, I open a private window with no extensions.

Here's the repo : https://gitlab.com/QuentinLeCaignec/some-text-adventure

This doesn't seem normal for the most basic example game, especially the tutorial game shown by the docs. Am I doing something wrong? Is this normal?

**Comment by eonarheim** [comment]
@QuentinLeCaignec That doesn't seem right at all! I don't think you are doing anything wrong, this should work as you expect.

I'm not able to reproduce this on my hardware using your sample repo.  (On my older Surface Book 2 I'm running 60fps pretty consistently)

![image](https://user-images.githubusercontent.com/612071/165349818-8a4ae43e-c7d5-42e7-8443-c9a67c173f99.png)

Can you collect some information for me:
1. What hardware are you running? CPU/Graphics etc
2. What OS are you running?
3. Are you running your device on battery?
4. Is your [chrome hardware acceleration](https://www.howtogeek.com/412738/how-to-turn-hardware-acceleration-on-and-off-in-chrome/#:~:text=In%20the%20Settings%20menu%2C%20expand,Relaunch%E2%80%9D%20to%20apply%20the%20changes.) setting currently on/off?
5. If at all possible can you collect a CPU profile from the chrome dev tools and share it with me?
  
![image](https://user-images.githubusercontent.com/612071/165350217-7055048a-f37b-4be3-88b2-4febefc5c527.png)

**Reply by MorganeLeCaignec** [reply]
Definitely seems hardware related. I'm not on the laptop I was testing this one yesterday but on my home desktop, the performances are different. On this new test, i'm on a linux desktop (Kubuntu, KDE Plasma 5.22, 5.13.13 kernel), with 16Go of RAM, an Nvidia GeForce GTX 1070 and a i7-8700 CPU.

On chrome, the performances are now smooth and stay very close to 60 with hardware acceleration on. With acceleration off, they're incredibly unstable, jumping up and down constantly.
On firefox the performances are just generally worse, even with hardware acceleration turned on. Firefox also considers hardware acceleration to not be a recommended setting, most people would not have it on, so ideally i'd like things to be stable on a non-hardware-accelerated browser.

Also, even with a steady 60 FPS, the collision ends up glitching at some point and the ball does something weird like phase through or "slide" against a surface in a strange way. Less likely when the FPS is stable, but still can happen, which doesn't seem normal to me.

So two follow-up questions :
Why are the performances so much worse on firefox compared to chrome, whether hardware acceleration is on or off? Is it only on my hardware?
Is hardware acceleration really necessary for such a basic game example? I'm doubting my ability to have games run well on multiple hardware and browsers if that's the case.

I will test on the laptop and send you details on the hardware when i'm able.

**Reply by eonarheim** [reply]
Hi @QuentinLeCaignec thanks for testing! I appreciate the feedback, and hardware information!

For the ball sliding on the surface issue you describe, the sample code isn't 100% the most robust way to handle this paddle/ball interaction. The issue stems from the handler reversing the velocity if a collision is detected, but if the ball is still in collision the next frame and hasn't left contact with the paddle, it will toggle velocity back and forth between frames effectively sticking it to the paddle. I'll update the example soon with a more robust handler that waits for ball/paddle separation before allowing another reversal.

For your first question: I think chrome and firefox do some hardware detection to turn on/off hardware acceleration so some hardware configurations might start with a un-ideal default (citation needed, I believe this is the case). Without hardware acceleration the WebGL based graphics need to be rendered using the CPU which is much much slower at drawing. 

Second, I totally agree that there should be a better non-accelerated experience for users, I'll look into some options. Definitely open to any suggestions you might have!

Some things I can think of for near term solutions:
* Notify users with a notification about hardware acceleration
* Analyze how other engines behave without hardware acceleration and their solutions
* Analyze profiler data on non-accelerated devices will illuminate any performance optimizations that can be made to the renderer

Some more long-term strategic ideas:
* Unhook the update portion of the mainloop and do work in a webworker freeing up the main UI thread for more render time
* Further optimize collision data structures which is still the biggest part of our frame budget
* Firefox garbage collection is especially unkind to excalibur, look into more pooling around Vector/BoundingBox types to avoid nursery (short lived object) garbage collections.

cc @kamranayub @jedeen Any ideas?

**Reply by eonarheim** [reply]
Another thing that might be interesting, if you toggle on the 2D Canvas fallback how do things perform for you?

```typescript
ex.Flags.useCanvasGraphicsContext();// enable before engine construction
new ex.Engine({...});
```

**Reply by eonarheim** [reply]
This (admittedly older thread) seems to indicate it's possible to drop to non-accelerated mode based on out of date graphics drivers. They also say that "There is no concrete plan for enabling GPU acceleration for 2D canvas on Linux due to continued concerns about stability." but this is also a very old thread.

chrome://gpu

https://groups.google.com/a/chromium.org/g/chromium-dev/c/09NnO6jYT6o

**Reply by kamranayub** [reply]
I would say it's reasonable for Excalibur to do the best we can given things under our control.

For things outside of our control, we are at the mercy of the browser and developers will need to make their own call for what they want (be okay with sub-par perf, show users troubleshooting info, require users to have hardware accel before continuing).

It would be good to know if this is common or an edge case. Yes we can do a lot of the things you mention @eonarheim but also need to keep in mind the impact / opportunity cost. If this is a problem faced by 1% of users, a dialog walking through how to enable hardware acceleration may be "good enough." 

Of course, much of what you mentioned would benefit everybody so still seems worth doing long term!

**Reply by MorganeLeCaignec** [reply]
@eonarheim I've tried enabling the Canvas fallback on Firefox and it seems more stable, less constant up/down but not as smooth as chrome. This is with or without hardware acceleration.

On chrome, enabling Canvas fallback with hardware acceleration doesn't seem to make a difference, the performances are smooth either way. Without hardware acceleration however, the Canvas fallback is noticeably more stable but still not perfectly smooth.

So if make a sort of ranking of stability from most stable to least stable :
- Chrome (HA) with WebGL / Chrome (HA) with Canvas
- Chrome (no HA) with Canvas
- Chrome (no HA) with WebGL / Firefox (HA) with Canvas
- Firefox (HA) with Webgl
- Firefox (no HA) with WebGL / Firefox (no HA) with Canvas

In general, Chrome can have spikes and sudden drops but seems smoother, whereas even with HA and Canvas firefox it never entirely stable.

I might have missed your answer, but do you know why Firefox seems to perform generally worse than chrome ?

**Reply by eonarheim** [reply]
@QuentinLeCaignec thanks for the thorough research!

Indeed, Firefox generally performs worse than chrome at least when it comes to graphics things.

**Reply by eonarheim** [reply]
@QuentinLeCaignec 

I've spent some time this week on #2309 and have managed to improve the performance on chrome/firefox a good deal. The non-HA is still not as good as I want, but I'm able to get the breakout sample to around 30fps. I think our near term solution (still in progress) will be an automatic 2D canvas fallback and/or a message to the user around enabling HA.

I also updated the sample to use more robust collision logic to prevent the sticking and oscillating problems you spotted https://github.com/excaliburjs/sample-breakout

I'm planning on a new version (v0.26.0) of Excalibur very soon (probably this weekend no promises though), but if you want to play with some of the performance bits early npm packages >= `excalibur@0.26.0-alpha.457` have the perf improvements.

**Reply by MorganeLeCaignec** [reply]
Wow! I wasn't expecting such rapid improvement. Well I'm glad if I helped a little with my feedback :)
I'll wait for the release, and I'll report back the changes on my hardware! Thank you!

**Reply by eonarheim** [reply]
We've merged this feature! https://github.com/excaliburjs/Excalibur/pull/2310
